File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/SuggestionsAreMade.java
package org.weaveit.seatingplacesuggestions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SuggestionsAreMade {
    private final String showId;
    private final int partyRequested;
    private final Map<PricingCategory, List<String>> seatsByCategory = new HashMap<>();

    public SuggestionsAreMade(String showId, int partyRequested) {
        this.showId = showId;
        this.partyRequested = partyRequested;
    }

    public void addSeats(PricingCategory category, List<String> seatNames) {
        seatsByCategory.computeIfAbsent(category, k -> new ArrayList<>()).addAll(seatNames);
    }

    public List<String> seatNames(PricingCategory pricingCategory) {
        return seatsByCategory.getOrDefault(pricingCategory, List.of());
    }

    public String showId() {
        return showId;
    }

    public int partyRequested() {
        return partyRequested;
    }
}

File: TheaterSuggestions/Kotlin/ScreeningSeatingArrangementSuggestions/src/main/kotlin/org/weaveit/seatingplacesuggestions/SuggestionsAreMade.kt
package org.weaveit.seatingplacesuggestions

open class SuggestionsAreMade(
    val showId: String,
    val partyRequested: Int,
    suggestions: List<SuggestionIsMade>
) {
    private val forCategory: Map<PricingCategory, List<SuggestionIsMade>> =
        organizeSuggestionsByCategory(suggestions)

    private fun organizeSuggestionsByCategory(
        suggestions: List<SuggestionIsMade>
    ): Map<PricingCategory, List<SuggestionIsMade>> {
        val result = mutableMapOf<PricingCategory, MutableList<SuggestionIsMade>>()

        for (category in PricingCategory.entries) {
            result[category] = mutableListOf()
        }

        for (suggestion in suggestions) {
            result[suggestion.pricingCategory]!!.add(suggestion)
        }

        return result.mapValues { it.value.toList() }
    }

    fun seatNames(pricingCategory: PricingCategory): List<String> {
        return forCategory[pricingCategory]!!
            .map { it.seatNames().joinToString("-") }
    }

    fun matchExpectations(): Boolean {
        return forCategory.values
            .flatten()
            .any { it.matchExpectation() }
    }
}
